django/contrib/admin/options.py

```diff
@@ -1650,7 +1651,14 @@ def changelist_view(self, request, extra_context=None):
         # Handle POSTed bulk-edit data.
         if request.method == 'POST' and cl.list_editable and '_save' in request.POST:
             FormSet = self.get_changelist_formset(request)
-            formset = cl.formset = FormSet(request.POST, request.FILES, queryset=self.get_queryset(request))
+            # Get the objects ids to filter the queryset to get only the objects that will be updated.
+            regexp = re.compile('{prefix}-(\d+)-id$'.format(prefix=FormSet.get_default_prefix()))
+            objects_id = []
+            for key, value in request.POST.items():
+                if regexp.match(key):
+                    objects_id.append(value)
+            formset = cl.formset = FormSet(request.POST, request.FILES,
+                                           queryset=self.get_queryset(request).filter(pk__in=objects_id))
```

Comment:
Can we use `cl.result_list` as in the next elif clause below? At any rate there should be a more direct way of getting the objects we're displaying on the page vs. using regex on POST items, which is too brittle and complex.

Reply:
Hi @akulakov,

If we use cl.result_list is possible that we get 500 errors because due to a concurrent problem (cl.result_list filter and order the datasets by the list view values but the datasets could change with a new object, deleting an object or modifying any order value) and for this problem was changed https://github.com/django/django/commit/917cc288a38f3c114a5440f0749b7e5e1086eb36#diff-3c42de3e53aba87b32c494f995a728dfR1583

Maybe the way to get the objects that will be updated is not the better possible but now is a great problem...

Sorry for take too long to respond.

Regards!

Reply:
Hey, if the problem is that the entire queryset is fetched to generate the formset - instead of fetching only the relevent objects using regexp, have you considered fetching just the current page instead? You have the paginator and you have the request. 

Obviously fetching only the neccesary objects to update is ideal but as @akulakov said, it's a bit brittle. Using the page is not the the best solution but it should contain the performance regression. 

Hope this makes sense :)

Reply:
Hi @hakib!

This was remove by this bug https://code.djangoproject.com/ticket/11313 because, althrougt you have the paginator and the request, the order or the queryset could be change so, for example, if you edit the last element of the list and any user, concurrntly, add a new object, the object that you edit does not appear in the page then the element will not be found (if you order the elements by id, for example).

Reply:
Hey @rubgombar1, thats right. Thanks for the reference.

I take comfort in the fact that I wasent the only one who didn't think about it :)